### PR TITLE
WIP: ELLIPTIC CURVE

### PR DESCRIPTION
**Pull request by abh2k**
revives #2449 

* ntheory
  * implemented  elliptic curve

### CODE REVIEW

**Comment by jksuom** [comment]
You should preserve the original commits in https://github.com/sympy/sympy/pull/2449. After fetching the PR, you should only add new commits to the same branch.

**Comment by abhinav28071999** [comment]
could you tell me how to fetch   that pr.

File: sympy/ntheory/ec.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+from sympy.core.numbers import oo
+from sympy.core.relational import Eq
+from sympy.polys.domains import FiniteField, QQ, RationalField
+from sympy.solvers.solvers import solve
+
+from .factor_ import divisors
+from .residue_ntheory import sqrt_mod
+
+
+class EllipticCurve():
```
Comment (czgdp1807) [inline sympy/ntheory/ec.py]: IMO, `()` can be removed.

File: sympy/ntheory/ec.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+    """
+    Create the following Elliptic Curve over domain.
+    `y^{2} + a_{1} x y + a_{3} y = x^{3} + a_{2} x^{2} + a_{4} x + a_{6}`
+    The default domain is ``QQ``. If no coefficient ``a1``, ``a2``, ``a3``,
+    it create curve as following form.
+    `y^{2} = x^{3} + a_{4} x + a_{6}`
+    Examples
```
Comment (czgdp1807) [inline sympy/ntheory/ec.py]: An empty line is needed here.

File: sympy/ntheory/ec.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    ========
+    References
```
Comment (czgdp1807) [inline sympy/ntheory/ec.py]: +1

File: sympy/ntheory/ec.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    ==========
+    [1] J. Silverman "A Friendly Introduction to Number Theory" Third Edition
+    [2] http://mathworld.wolfram.com/EllipticDiscriminant.html
+    [3] G. Hardy, E. Wright "An Introduction to the Theory of Numbers" Sixth Edition
+    """
+
+    def __init__(self, a4, a6, a1=0, a2=0, a3=0, domain=QQ):
```
Comment (czgdp1807) [inline sympy/ntheory/ec.py]: Generally, in SymPy, `__new__` is used.

File: sympy/ntheory/ec.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+        self._a3 = self._domain(a3)
+        self._a4 = self._domain(a4)
+        self._a6 = self._domain(a6)
+        self._eq = Eq(y**2 + self._a1*x*y + self._a3*y, x**3 + self._a2*x**2 + self._a4*x + self._a6)
+        if isinstance(self._domain, FiniteField):
+            self._rank = 0
+        elif isinstance(self._domain, RationalField):
+            self._rank = None
+
+    def __call__(self, x, y, z=1):
```
Comment (czgdp1807) [inline sympy/ntheory/ec.py]: Why do we need this?

File: sympy/ntheory/ec.py [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+            return True
+        return self._eq.subs({x: x1, y: y1})
+
+    def __repr__(self):
+        return 'E({}): {}'.format(self._domain, self._eq)
+
+    def minimal(self):
+        """
+        Return minimal Weierstrass equation.
+        Examples
```
Comment (czgdp1807) [inline sympy/ntheory/ec.py]: > An empty line is needed here.

+1

File: sympy/ntheory/ec.py [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+        ========
+        >>> from sympy.ntheory.ec import EllipticCurve
```
Comment (czgdp1807) [inline sympy/ntheory/ec.py]: +1

**Comment by jksuom** [comment]
I think that `git fetch origin pull/2449/head:elliptic-curve` should create a new branch with the original commits.

File: sympy/ntheory/ec.py [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+        p = self * 2
+        if p.y == -self.y:  # 2P = -P
+            return 3
+        i = 2
+        while int(p.x) == p.x:
+            p = self + p
+            i += 1
+            if p.z == 0:
+                return i
+        return oo
```
Comment (czgdp1807) [inline sympy/ntheory/ec.py]: A new line is needed at the end.

**Pull request closed without merging**